File: backend/controllers/authController.js
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import db from '../config/db.js';

const JWT_SECRET = process.env.JWT_SECRET || 'fallback_super_secret_key_123';

// Register User
export const register = async (req, res) => {
    try {
        const { fullName, email, password, role, degree } = req.body;

        // Validation
        if (!fullName || !email || !password || !role) {
            return res.status(400).json({ message: 'Missing required fields.' });
        }

        // Check if user already exists
        db.get('SELECT email FROM users WHERE email = ?', [email], async (err, row) => {
            if (err) return res.status(500).json({ message: 'Database error', error: err.message });
            if (row) return res.status(409).json({ message: 'Email already in use.' });

            // Hash password
            const salt = await bcrypt.genSalt(10);
            const hashedPassword = await bcrypt.hash(password, salt);

            // Handle optional files
            const resumeFilePath = req.files && req.files['resume'] ? req.files['resume'][0].path.replace(/\\/g, '/') : null;
            const studentIdFilePath = req.files && req.files['studentId'] ? req.files['studentId'][0].path.replace(/\\/g, '/') : null;

            // Insert new user
            const verificationStatus = 'PENDING';
            const sql = `INSERT INTO users (fullName, email, password, role, degree, resumeFilePath, studentIdFilePath, verification_status)
                         VALUES (?, ?, ?, ?, ?, ?, ?, ?)`;

            db.run(sql, [fullName, email, hashedPassword, role, degree || null, resumeFilePath, studentIdFilePath, verificationStatus], function (err) {
                if (err) return res.status(500).json({ message: 'Error registering user', error: err.message });

                res.status(201).json({
                    message: 'User registered successfully!',
                    user: { id: this.lastID, fullName, email, role }
                });
            });
        });
    } catch (error) {
        res.status(500).json({ message: 'Server error during registration.', error: error.message });
    }
};

// Login User
export const login = (req, res) => {
    try {
        const { email, password } = req.body;

        if (!email || !password) {
            return res.status(400).json({ message: 'Please provide email and password.' });
        }

        // Find user
        db.get('SELECT * FROM users WHERE email = ?', [email], async (err, user) => {
            if (err) return res.status(500).json({ message: 'Database error', error: err.message });
            if (!user) return res.status(404).json({ message: 'Invalid credentials. User not found.' });

            // Verify password
            const isMatch = await bcrypt.compare(password, user.password);
            if (!isMatch) return res.status(401).json({ message: 'Invalid credentials. Incorrect password.' });

            // Verification Check for All Users
            if (user.verification_status === 'PENDING') {
                return res.status(403).json({ message: 'Verifying your data' });
            }
            if (user.verification_status === 'REJECTED') {
                return res.status(403).json({ message: 'Unfortunately your data not matched our standards' });
            }

            // Generate JWT
            const token = jwt.sign(
                { id: user.id, email: user.email, role: user.role },
                JWT_SECRET,
                { expiresIn: '2h' }
            );

            res.status(200).json({
                message: 'Login successful!',
                token,
                user: {
                    id: user.id,
                    fullName: user.fullName,
                    email: user.email,
                    role: user.role
                }
            });
        });

    } catch (error) {
        res.status(500).json({ message: 'Server error during login.', error: error.message });
    }
};

// Reset Password
export const resetPassword = async (req, res) => {
    try {
        const { email, newPassword } = req.body;

        if (!email || !newPassword) {
            return res.status(400).json({ message: 'Please provide email and new password.' });
        }

        // Find user
        db.get('SELECT * FROM users WHERE email = ?', [email], async (err, user) => {
            if (err) return res.status(500).json({ message: 'Database error', error: err.message });
            if (!user) return res.status(404).json({ message: 'User not found with that email address.' });

            // Hash new password
            const salt = await bcrypt.genSalt(10);
            const hashedPassword = await bcrypt.hash(newPassword, salt);

            // Update user's password
            db.run('UPDATE users SET password = ? WHERE id = ?', [hashedPassword, user.id], function (err) {
                if (err) return res.status(500).json({ message: 'Error updating password', error: err.message });

                res.status(200).json({ message: 'Password reset successfully!' });
            });
        });
    } catch (error) {
        res.status(500).json({ message: 'Server error during password reset.', error: error.message });
    }
};

// Delete Own Account
export const deleteOwnAccount = (req, res) => {
    try {
        const { id } = req.params;

        // Check if user exists
        db.get('SELECT * FROM users WHERE id = ?', [id], (err, user) => {
            if (err) return res.status(500).json({ message: 'Database error', error: err.message });
            if (!user) return res.status(404).json({ message: 'User not found.' });

            // Delete user
            db.run('DELETE FROM users WHERE id = ?', [id], function (err) {
                if (err) return res.status(500).json({ message: 'Error deleting account', error: err.message });

                res.status(200).json({ message: 'Account deleted successfully.' });
            });
        });
    } catch (error) {
        res.status(500).json({ message: 'Server error during account deletion.', error: error.message });
    }
};

File: backend-java/src/main/java/com/careerguidance/controller/AuthController.java
package com.careerguidance.controller;

import com.careerguidance.dto.AuthResponse;
import com.careerguidance.dto.LoginRequest;
import com.careerguidance.dto.ResetPasswordRequest;
import com.careerguidance.model.User;
import com.careerguidance.repository.UserRepository;
import com.careerguidance.security.JwtUtils;
import com.careerguidance.service.FileStorageService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.security.core.context.SecurityContextHolder;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    @Autowired
    AuthenticationManager authenticationManager;

    @Autowired
    UserRepository userRepository;

    @Autowired
    PasswordEncoder encoder;

    @Autowired
    JwtUtils jwtUtils;

    @Autowired
    FileStorageService fileStorageService;

    @PostMapping("/register")
    public ResponseEntity<?> registerUser(@RequestParam("fullName") String fullName,
                                          @RequestParam("email") String email,
                                          @RequestParam("password") String password,
                                          @RequestParam("role") String role,
                                          @RequestParam(value = "degree", required = false) String degree,
                                          @RequestParam(value = "resume", required = false) MultipartFile resume,
                                          @RequestParam(value = "studentId", required = false) MultipartFile studentId) {
        try {
            if (userRepository.findByEmail(email).isPresent()) {
                Map<String, String> response = new HashMap<>();
                response.put("message", "Email already in use.");
                return ResponseEntity.status(HttpStatus.CONFLICT).body(response);
            }

            User user = new User();
            user.setFullName(fullName);
            user.setEmail(email);
            user.setPassword(encoder.encode(password));
            user.setRole(role);
            user.setDegree(degree);
            user.setVerificationStatus("PENDING");

            if (resume != null && !resume.isEmpty()) {
                String resumePath = fileStorageService.saveFile(resume, "resumes");
                user.setResumeFilePath(resumePath);
            }

            if (studentId != null && !studentId.isEmpty()) {
                String studentIdPath = fileStorageService.saveFile(studentId, "studentIds");
                user.setStudentIdFilePath(studentIdPath);
            }

            userRepository.save(user);

            Map<String, Object> response = new HashMap<>();
            response.put("message", "User registered successfully!");
            response.put("user", AuthResponse.UserResponse.fromUser(user));
            return ResponseEntity.status(HttpStatus.CREATED).body(response);
        } catch (Exception e) {
            Map<String, String> response = new HashMap<>();
            response.put("message", "Server error during registration.");
            response.put("error", e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response);
        }
    }

    @PostMapping("/login")
    public ResponseEntity<?> authenticateUser(@RequestBody LoginRequest loginRequest) {
        try {
            Optional<User> userOpt = userRepository.findByEmail(loginRequest.getEmail());
            if (userOpt.isEmpty()) {
                Map<String, String> response = new HashMap<>();
                response.put("message", "Invalid credentials. User not found.");
                return ResponseEntity.status(HttpStatus.NOT_FOUND).body(response);
            }

            User user = userOpt.get();

            if (user.getVerificationStatus().equals("PENDING")) {
                Map<String, String> response = new HashMap<>();
                response.put("message", "Verifying your data");
                return ResponseEntity.status(HttpStatus.FORBIDDEN).body(response);
            }
            if (user.getVerificationStatus().equals("REJECTED")) {
                Map<String, String> response = new HashMap<>();
                response.put("message", "Unfortunately your data not matched our standards");
                return ResponseEntity.status(HttpStatus.FORBIDDEN).body(response);
            }

            Authentication authentication = authenticationManager.authenticate(
                    new UsernamePasswordAuthenticationToken(loginRequest.getEmail(), loginRequest.getPassword()));

            SecurityContextHolder.getContext().setAuthentication(authentication);
            String jwt = jwtUtils.generateJwtToken(authentication);

            AuthResponse response = new AuthResponse("Login successful!", jwt, AuthResponse.UserResponse.fromUser(user));
            return ResponseEntity.ok(response);
        } catch (org.springframework.security.authentication.BadCredentialsException e) {
            Map<String, String> response = new HashMap<>();
            response.put("message", "Invalid credentials. Incorrect password.");
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(response);
        } catch (Exception e) {
            Map<String, String> response = new HashMap<>();
            response.put("message", "Server error during login.");
            response.put("error", e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response);
        }
    }

    @PostMapping("/reset-password")
    public ResponseEntity<?> resetPassword(@RequestBody ResetPasswordRequest request) {
        Optional<User> userOpt = userRepository.findByEmail(request.getEmail());
        if (userOpt.isEmpty()) {
            Map<String, String> response = new HashMap<>();
            response.put("message", "User not found with that email address.");
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(response);
        }

        User user = userOpt.get();
        user.setPassword(encoder.encode(request.getNewPassword()));
        userRepository.save(user);

        Map<String, String> response = new HashMap<>();
        response.put("message", "Password reset successfully!");
        return ResponseEntity.ok(response);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteOwnAccount(@PathVariable Long id) {
        try {
            if (!userRepository.existsById(id)) {
                Map<String, String> response = new HashMap<>();
                response.put("message", "User not found.");
                return ResponseEntity.status(HttpStatus.NOT_FOUND).body(response);
            }
            userRepository.deleteById(id);
            Map<String, String> response = new HashMap<>();
            response.put("message", "Account deleted successfully.");
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            Map<String, String> response = new HashMap<>();
            response.put("message", "Server error during account deletion.");
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response);
        }
    }
}
